### PR TITLE
Custom Branding

### LINKED ISSUE #725: [Proposal] Custom Branding
### Problem Statement
One of the customers’ pain points is to not be able to easily customize OpenSearch Dashboards. If a customer wants dashboard customization, they need to manually fork and change the source code. However, changing source code takes too much time and is likely to fail after each upgrade; on the other hand, lack of engineering qualifications and resources also pose a great challenge to this issue.
We are proposing a way for users to customize OpenSearch Dashboard using opensearch_dashboards.yml file and starting with predefined places such as title and logo. Users here are defined as the cluster administrator who configure the application and has access to the opensearch_dashboards.yml file before it is hosted. 

### Requirements
Users can configure logo and application title using opensearch_dashboards.yml file. Detailed tutorial and documentation will be provided for users and developers to follow. 

### Implementation Proposal
This approach introduces additional configurations to configuration file opensearch_dashboard.yml. The configs will be added in opensearch_dashboards config and will be directly passed and used in core. Classes in core will have access to the configs and render the Dashboard UI based on what user provided in the configuration file. The proposed added configurations are: 

* opensearchDashboards.branding.custom_enabled: [boolean]
* opensearchDashboards.branding.logoURL:  [string]
* opensearchDashboards.branding.logoDarkmodeURL: [string]
* opensearchDashboards.branding.title: [string]
* opensearchDashboards.branding.loadingLogoURL: [string]
* opensearchDashboards.branding.loadingLogoDarkmodeURL: [string]

### PR DESCRIPTION
**Pull request by abbyhu2000**
### Description
This PR is proposing a way for users to customize OpenSearch Dashboard using opensearch_dashboards.yml file and starting with predefined list of titles and logos. Our initial target users will be the cluster administrator who has access to the opensearch_dashboards.yml file and who is able to configure the application before it is hosted. This is a starting point and foundation for a bigger effort of branding and theming configurations in the future.

### Added Configurations in opensearch_dashboards.yml file
Screen Shot 2021-09-28 at 11 07 12 AM

### Targeted Logo and Title(7 logos and 4 titles)
Screen Shot 2021-09-28 at 11 09 50 AM

### Config Details
1. mark.defaultUrl (L1 - L6): simple one step branding configurations; if no other config are provided, a valid URL for mark.defaultUrl config will be rendered all 6 logo occurrences; if not provided, original opensearch logos will be rendered
2. logo.defaultUrl (L5): added option for user to display a fuller logo on the navigation bar; if not provided, fall back to mark.defaultUrl
3. loadingLogo.defaultUrl (L2): added option for user to display a custom loading logo on the loading page; if provided, render the custom loading logo; if not, fall back to mark.defaultUrl and render the static mark with a loading bar
4. mark.darkModeUrl, logo.darkModeUrl, loadingLogo.darkModeUrl: added option for user to configure logo in both default settings and dark mode settings; if darkModeUrl not provided, fall back to defaultUrl
5. faviconUrl (L7): basic browser favicon configuration(Chrome, Firefox, Edge and Safari); advanced favicon configurations(Windows Title, Android Device icon) are not supported 
6. applicationTitle (T1 - T4): replace "OpenSearch Dashboards" with a custom title

### URL requirements
1. For image URLs, only svg, png, and gif file types are allowed; any image size is permitted, but will be formatted to certain size for best appearance
2. If user inputs invalid URL, log will display a info message
3. If user inputs dark mode URLs without providing a valid default mode URLs, log will display a error message; dark mode URLs can only be considered after having valid default mode URLs 
4. For config applicationTitle, length of the input string needs to be between 1 and 36 letters.

Signed-off-by: Abby Hu abigailhu2000@gmail.com

### Issues Resolved
#725 
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by abbyhu2000** [comment]
Screen Shot 2021-09-28 at 3 58 51 PM

**Review by kavilla — DISMISSED** [review]

**Review by ananzh — DISMISSED** [review]
nice job Abby

**Review by boktorbb-amzn — DISMISSED** [review]
Very nice work!

**Review by kgcreative — DISMISSED** [review]
Great work!

**Review by tmarkley — DISMISSED** [review]
Great work! I think we can reduce some duplication here and clean up the comments/documentation.

**Review by kavilla — APPROVED** [review]

File: src/core/public/chrome/ui/header/collapsible_nav.tsx [hunk 1]
```diff
@@ -137,6 +140,42 @@ export function CollapsibleNav({
...
+  const DEFAULT_OPENSEARCH_MARK = `${branding.assetFolderUrl}/opensearch_mark_default_mode.svg`;
+  const DARKMODE_OPENSEARCH_MARK = `${branding.assetFolderUrl}/opensearch_mark_dark_mode.svg`;
+
+  const darkMode = branding.darkMode;
+  const markDefault = branding.mark?.defaultUrl;
+  const markDarkMode = branding.mark?.darkModeUrl;
+
+  /**
+   * Use branding configurations to check which URL to use for rendering
+   * side menu opensearch logo in default mode
```
Comment (tmarkley) [inline src/core/public/chrome/ui/header/collapsible_nav.tsx:152]: Note for follow-up changes: use correct cases for `OpenSearch`

**Review by tmarkley — APPROVED** [review]
@abbyhu2000 great work, and @kavilla thanks for these updates!

**Pull request merged**